metaserver/src/main/scala/scala/meta/languageserver/search/SymbolIndex.scala

```diff
@@ -220,47 +267,81 @@ class SymbolIndex(
     }
   }
 
-  /** Returns a list of fallback symbols that can act instead of given symbol. */
-  private def alternatives(symbol: Symbol): List[Symbol] =
-    symbol match {
+  // TODO(alexey) this is not index specific and could be moved somewhere else (like Symbol-ops)
+  case object alternatives {
+
+    /** Returns a list of fallback symbols that can act instead of given symbol. */
+    // TODO(alexey) review this list
+    def forReferences(symbol: Symbol): List[Symbol] = {
+      List(
+        caseClassCompanionToType,
+        caseClassApplyOrCopyParams
+      ).flatMap { _.lift(symbol) }
+    }
+
+    /** Returns a list of fallback symbols that can act instead of given symbol. */
+    def forDefinition(symbol: Symbol): List[Symbol] = {
+      List(
+        caseClassCompanionToType,
+        caseClassApplyOrCopy,
+        caseClassApplyOrCopyParams,
+        methodToTerm
+      ).flatMap { _.lift(symbol) }
+    }
```

Comment:
There's an issue with it:

```
case class User(name: String, age: Int)

object User {
  def apply(name: String): User = <<User>>(name, 2)
}
```

Go-to-definition on the selected `User(name, 2)` will redirect to the companion object. Can we workaround it somehow?

Reply:
Yes we can work around that but it will require a bit of work :) what we need to do is to refine the `indexDocument` method to look into the "Synthetics" section for `.apply`. I propose we get this PR merged in as soon as we have tests covering basic cases and then we refine the details in future PRs.